nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/src/main/java/org/apache/nifi/amqp/processors/AMQPPublisher.java

```diff
@@ -0,0 +1,128 @@
+
+import com.rabbitmq.client.AMQP.BasicProperties;
+import com.rabbitmq.client.Connection;
+import com.rabbitmq.client.ReturnListener;
+
+/**
+ * Generic publisher of messages to AMQP-based messaging system. It is based on
+ * RabbitMQ client API (https://www.rabbitmq.com/api-guide.html)
+ */
+final class AMQPPublisher extends AMQPWorker {
+
+    private final static Logger logger = LoggerFactory.getLogger(AMQPPublisher.class);
+
+    private final String exchangeName;
+
+    private final String routingKey;
+
+    /**
+     * Creates an instance of this publisher
+     *
+     * @param connection
+     *            instance of AMQP {@link Connection}
+     * @param exchangeName
+     *            the name of AMQP exchange to which messages will be published.
+     *            If not provided 'default' exchange will be used.
+     * @param routingKey
+     *            (required) the name of the routingKey to be used by AMQP-based
+     *            system to route messages to its final destination (queue).
+     */
+    AMQPPublisher(Connection connection, String exchangeName, String routingKey) {
+        super(connection);
+        this.validateStringProperty("routingKey", routingKey);
+        this.exchangeName = exchangeName == null ? "" : exchangeName.trim();
+        if (this.exchangeName.length() == 0) {
+            logger.info("The 'exchangeName' is not specified. Messages will be sent to default exchange");
+        }
+
+        this.routingKey = routingKey;
+        this.channel.addReturnListener(new UndeliverableMessageLogger());
+        logger.info("Successfully connected AMQPPublisher to " + connection.toString() + " and '" + this.exchangeName
+                + "' exchange with '" + routingKey + "' as a routing key.");
+    }
+
+    /**
+     * Publishes message without any AMQP properties (see
+     * {@link BasicProperties}) to a pre-defined AMQP Exchange.
+     *
+     * @param bytes
+     *            bytes representing a message.
+     */
+    void publish(byte[] bytes) {
+        this.publish(bytes, null);
+    }
+
+    /**
+     * Publishes message with provided AMQP properties (see
+     * {@link BasicProperties}) to a pre-defined AMQP Exchange.
+     *
+     * @param bytes
+     *            bytes representing a message.
+     * @param properties
+     *            instance of {@link BasicProperties}
+     */
+    void publish(byte[] bytes, BasicProperties properties) {
+        if (this.channel.isOpen()) {
+            try {
+                this.channel.basicPublish(this.exchangeName, this.routingKey, true, properties, bytes);
+            } catch (Exception e) {
+                throw new IllegalStateException("Failed to publish to '" +
+                        this.exchangeName + "' with '" + this.routingKey + "'.", e);
+            }
+        } else {
+            throw new IllegalStateException("This instance of AMQPPublisher is invalid since "
+                    + "its publishigChannel is closed");
```

Comment:
typo "publishingChannel". Also should this throw ClosedChannelException rather than IllegalStateException?